compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -975,20 +971,21 @@ public void visit(BLangSimpleVariable varNode, AnalyzerData data) {
         }
         SymbolEnv currentEnv = data.env;
         int ownerSymTag = currentEnv.scope.owner.tag;
-        boolean isListenerDecl = varNode.flagSet.contains(Flag.LISTENER);
-        if ((ownerSymTag & SymTag.INVOKABLE) == SymTag.INVOKABLE || (ownerSymTag & SymTag.LET) == SymTag.LET
-                || currentEnv.node.getKind() == NodeKind.LET_CLAUSE) {
+        Set<Flag> flagSet = varNode.flagSet;
+        boolean isListenerDecl = flagSet.contains(Flag.LISTENER);
+        if (flagSet.contains(Flag.OBJECT_FIELD)) {
+            analyzeVarNode(varNode, data, AttachPoint.Point.OBJECT_FIELD, AttachPoint.Point.FIELD);
+        } else if (flagSet.contains(Flag.RECORD_FIELD)) {
```

Comment:
Can't we use the flag `FIELD` instead of defining new flags. If we set env correctly then we can use `ownerSymTag`  to separate the record field and class field. right?

Reply:
What do you mean by setting env correctly? 
I think that it is clean to having to type for field.

Reply:
With this change when visiting a field of a class, the env is set to packageEnv. IMO it should be classEnv.
If we are going to introduce two new flags for each type then we can remove the `FIELD` flag. right?

Reply:
I don't think we need to change the env when visiting fields. Fields without self should not be seen within the class.

Reply:
There are few usages of `FIELD`, therefore we cannot use that

Reply:
The field should still be defined in the scope of the relevant symbol. So why can't we rely on FIELD + owner symbol? We shouldn't have to define new flags.

Reply:
Since we are using an already set scope as mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/36532#issuecomment-1254848680, the owner is not the object anymore, it is the owner of the object. That is why I followed this approach.

Reply:
That is wrong, right? The field's owner should be the object. We just need to make sure the fields aren't visible when visiting the default value expression.

Reply:
I think it is not okay to say that the `currentEnv.scope.owner` gives the owner of field. It is what the scopes owner. 
Similar to fields that I have done this PR, for methods also we are not using a new scope, it uses the scope what is already set. Please check https://github.com/ballerina-platform/ballerina-lang/blob/26592f29cd6d0889f4ad47f85df695341257c532/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java#L700

Reply:
I'm not sure why that was done for methods, I think it could be a different env but still a class env, we can check maybe. But, since we've been defining fields in the proper (or most relevant scope) why would be go back and define it in a less accurate (wider) scope?

The issue you are trying to solve only applies to the default value expression. I would rather we use a correct env (where the other fields are not visible, maybe somewhat similar to what was done in https://github.com/ballerina-platform/ballerina-lang/pull/35326) than change how the field itself is defined.

Reply:
@MaryamZi, the approach is changed, please have a look.